Add proper EMP behavior to TS aircraft

reaperrr opened the pull request:
This PR
- makes `Aircraft` pausable-conditional
- makes landed, EMP'd TS aircraft behave like EMP'd ground vehicles
- makes airborne TS aircraft crash when EMP'd

Closes #17096.
Fixes #16421.

PunkPun committed:
Make EMP effect carryall and orca transport

reaperrr committed:
Make Aircraft PausableConditional

Note: This commit only does the minimum changes
to implement PausableConditional, there are no
logic changes yet (like disabling movement on PauseOnCondition).

reaperrr committed:
Make Aircraft actually pausable/disableable

This commit makes aircraft
- ignore any aircraft-specific orders while disabled
- show blocked move cursor while paused
- set speeds to zero while paused or disabled

reaperrr committed:
Make TS aircraft crash when EMP'd mid-air

And refuse move orders if EMP'd while landed.
Matches original behaviour.

tovl commented:
Really nice!

Some things I noticed:
- Airborne units don't get the visual overlay when EMP-ed
- Landed units can still be ordered to take off, but they will crash instantly.

reaperrr commented:
> Airborne units don't get the visual overlay when EMP-ed

This will be a bit tricky to fix, since the tumbling aircraft husk is not the original aircraft, and we can't be sure that its spawner wasn't shot down by normal means. I'd like to leave that aspect for later.

> Landed units can still be ordered to take off, but they will crash instantly.

I'll look into that.

PunkPun reviewed (commented):
I did some testing and noticed these
- When a force landed carryall with cargo is EMPed it crashes
- Infantry can enter / exit Orka transport when it's EMPed
- EMP doesn't pause the rearming process

GraionDilach commented:
> > Airborne units don't get the visual overlay when EMP-ed
> 
> This will be a bit tricky to fix, since the tumbling aircraft husk is not the original aircraft, and we can't be sure that its spawner wasn't shot down by normal means. I'd like to leave that aspect for later.

Can't you base something on the framework which transfers the IronCurtain in RA from MCV to conyard?

pchote commented:
It should be as simple as adding
```
TransferTimedExternalConditionOnTransform:
	Condition: empdisable
```

to the aircraft.

pchote commented:
From https://github.com/OpenRA/OpenRA/pull/17189#issuecomment-538766314, IMO it would be best if we could make `Aircraft` pausing replace `LandOnCondition` (and fix it so it actually works). Is that too much of a scope creep here?

reaperrr commented:
> It should be as simple as adding `TransferTimedExternalConditionOnTransform` to the aircraft

Unfortunately it isn't, we need an equivalent for `SpawnActorOnDeath`, but that needs an equivalent of `void INotifyTransform.AfterTransform(Actor toActor)` first.

> From #17189 (comment), IMO it would be best if we could make Aircraft pausing replace LandOnCondition (and fix it so it actually works). Is that too much of a scope creep here?

Well I can try, might still shout for help or a deferral afterwards, though.

reaperrr committed:
Display EMP effect on EMP'd aircraft's husks

When EMP'd while airborne.

reaperrr committed:
Fix Harpy husk rotor offset

reaperrr committed:
Refactor Spin to MaximumSpinSpeed

Additionally, add descriptions to
other FallsToEarth properties.

reaperrr committed:
Replace LandOnCondition with landing when paused

reaperrr commented:
Rebased and updated.

- Enabled emp effect on husks by spawning a different husk when killed by EMP
- Fixed Harpy husk rotor offset
- Refactored `FallsToEarth` spin logic (the old one made aircraft spin way too fast)
- Removed `LandOnCondition` in favor of pausing, remove/comment out `KillsSelf` on any EMP-able TS aircraft to test

GraionDilach commented:
> Refactored FallsToEarth spin logic (the old one made aircraft spin way too fast)

A much more lenient option - even your setup is too fast for RA2 Kirov.
https://github.com/AttacqueSuperior/Engine/blob/master/OpenRA.Mods.AS/Traits/Air/FallsToEarthAS.cs#L25

tovl commented:
- Landed EMP-ed aircraft still instantly crash when ordered to take off instead of staying put in the latest version.
- Landed aircraft get darkened when they are EMP-ed while airborne aircraft do not. They do get the proper overlay though.

>  - Infantry can enter / exit Orka transport when it's EMPed
> - EMP doesn't pause the rearming process

These issues are also still present. The former is also the case for APCs, so that might be out of scope here.

reaperrr commented:
> Landed aircraft get darkened when they are EMP-ed while airborne aircraft do not. They do get the proper overlay though

I know it's technically inconsistent, but to me, the darkening has always mainly served as "you can't use this right now", indicating this will resume working after some time, while husks are mainly cosmetics as the original aircraft is already dead at that point.

tovl commented:
Yes that makes perfect sense if you view the darkening as purely non-diegetic. I hadn't really looked at it that way.

reaperrr commented:
Rebased and updated.

> Landed EMP-ed aircraft still instantly crash when ordered to take off instead of staying put in the latest version.

Fixed in c316111 (had to add a work-around to `Fly` activity, properly refactoring aircraft activities is out of scope for this PR).

> - Infantry can enter / exit Orka transport when it's EMPed
> - EMP doesn't pause the rearming process

I'm afraid both of these are out of scope for this PR.
`Cargo` isn't pausable-conditional (not conditional at all, actually), while the rearm part depends on my still-incomplete resupply refactor roadmap. 
Reloading ammo pools will eventually shift to `ReloadAmmoPool` entirely, but the minimum prerequisite for that is to grant a condition during resupply, and that's something I'd like to leave for a follow-up.

pchote reviewed (commented):
LGTM overall, great job!

Just a couple of minor issues and comments.

I do still think that paused aircraft above invalid terrain should seek out a valid landing location (to match the behaviour of subterranean units), but that is outside the scope of this PR.

reaperrr commented on mods/ts/rules/husks.yaml:
```diff
@@ -21,6 +21,12 @@ ORCA.Husk:
 
+ORCA.Husk.EMP:
+	Inherits: ORCA.Husk
+	WithIdleOverlay@EMPDISABLE:
```
Split `^EmpVisualEffects` from `^EmpDisable`, so they can be used by husks directly.

reaperrr commented:
Updated, but this broke husk spawning and I have no idea why yet.

pchote commented:
> Updated, but this broke husk spawning and I have no idea why yet.

This happens because `KillsSelf` kills the actor directly inside its `TraitEnabled`, which means that `INotifyKilled.Killed` runs in the middle of the `empdisabled` condition change notification. 

The rule reorganization means that the non-EMP `SpawnActorOnDeath` is notified first and disables itself. `KillsSelf`, then runs and kills the actor `SpawnActorOnDeath@EMP` is notified of the death and, because doesn't know about the condition change, is still disabled.

A new commit that wraps `KillsSelf`'s `Kill(self)` in a frame end task fixes the issue.

reaperrr committed:
Fix trait order issues with KillsSelf

reaperrr commented:
Updated.

pchote reviewed (commented):
Found another regression while testing. Fortunately, its an easy fix. Or if you'd rather get everything else merged asap I guess we could push 6eb9444dbdd92b905003674aa18d164cdfa8c003 into a different PR?

reaperrr commented:
Updated. Tested both RA and TS, and both looked good to me.

pchote commented on OpenRA.Mods.Common/Activities/Air/FallToEarth.cs:
```diff
@@ -20,16 +20,14 @@ public class FallToEarth : Activity
 			IsInterruptible = false;
 			aircraft = self.Trait<Aircraft>();
-			if (info.Spins)
-				acceleration = self.World.SharedRandom.Next(2) * 2 - 1;
```
I just noticed that this is removing the random spin direction - all units will now spin anti-clockwise.

pchote commented on OpenRA.Mods.Common/Traits/Air/FallsToEarth.cs:
```diff
@@ -19,10 +19,17 @@ namespace OpenRA.Mods.Common.Traits
 		public readonly string Explosion = "UnitExplode";
 
-		public readonly bool Spins = true;
+		[Desc("How fast the tumbling aircraft spins (at most). Negative values default to 'unlimited.",
```
Typo: `'unlimited.'.

Minor nit, but since this needs to be changed anyway: putting the value in quotes IMO feels like it is implying that 'unlimited' is a valid value.

How about
`[Desc("Limit the maximum spin (in facing units per tick) that can be achieved while crashing. 0 disables spinning. Negative values imply no limit.")]`?

reaperrr commented:
Updated, restored random spin direction (tested in TS) and fixed description.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/cef514d073fb982cc7697c729787c1cca0bbec1b
